### ISSUE TITLE
command-exists does not work for "cd DIRNAME" on Windows

### ISSUE THREAD

**Issue by KSR4599** [hapi-server/server-nodejs#24]
In the process of resolving [#22](https://github.com/hapi-server/server-nodejs/issues/22) where I had to use the module [command-exists](https://www.npmjs.com/package/command-exists) inorder to determine whether the "catalog" is a URL, file(path) or a command , I have encountered that the behavior of the module (In [metadata.js](https://github.com/hapi-server/server-nodejs/blob/4bf734b8e00632f81b73ab9a3653944959f8c4a4/lib/metadata.js#L245)) does not seem to be accurate and uniform. In particular, it is failing to determine ['cd'](https://github.com/hapi-server/server-nodejs/blob/master/metadata/SSCWeb.json#L16) as a command in windows (But works in Ubuntu). Moreover, it seems the path to the files (for [example](https://github.com/hapi-server/server-nodejs/blob/313dc05684b36d2655bd76bdd9db91b625c5ec2f/metadata/Example3.json#L26)) are also being considered as commands. 
I have done some unit testing :
```
var commandExistsSync = require('command-exists').sync;

var cmd = 'cd';
var cmd1 = 'C:\\Users\\ksred\\Desktop\\test\\server-nodejs-master/metadata/Example3/dataset1_info.json';
//In case of Unix, path of file would be something like:
//var cmd1 = '/home/srikar/real-hapi/server-nodejs-master/metadata/Example3/dataset1_info.json';

console.log("Check for cd : "+ commandExistsSync(cmd));

console.log("Check for Path : "+ commandExistsSync(cmd1));
```

The output in Windows : 

> Check for cd : false
> Check for Path : true

The output in Ubuntu : 

> Check for cd : true
> Check for Path : true

**Comment by rweigel** [comment]
Is `cd` alone a valid command in Windows?

Try this on Windows and Ubuntu:

```
var cmdExistsSync = require('command-exists').sync;
var fileExistsSync = require('fs').existsSync;

var cmd = 'cd';
var file = 'server.js';

//OS-X node 6.17.1
//cmdExistsSync('cd'):         true
//cmdExistsSync('server.js'):  false
//fileExistsSync('cd'):        false
//fileExistsSync('server.js'): true
```

If `cd` is a valid command on Windows, would it work to check if it is a file first and if not, assume a command?

**Comment by KSR4599** [comment]
Windows Output :
```
cmdExistsSync('cd'):         false
cmdExistsSync('server.js'):  true
fileExistsSync('cd'):        false
fileExistsSync('server.js'): true
```
Ubuntu Output:
```
cmdExistsSync('cd'):         true
cmdExistsSync('server.js'):  false
fileExistsSync('cd'):        false
fileExistsSync('server.js'): true
```
After seeing the above outputs, seems like checking if it is a file first and if not, finally assuming it as a command looks like a better option.

**Comment by rweigel** [comment]
Is `cd` a valid command in Windows?

**Comment by KSR4599** [comment]
Yes, `cd` is a valid command in windows. I am assuming the issue is with the command-exists module to treat it like a command.

**Comment by KSR4599** [comment]
Yes, `cd` is a valid command in windows. I am assuming the issue is with the command-exists module to treat it like a command.

**Comment by rweigel** [comment]
OK. Please post an issue at https://github.com/mathisonian/command-exists/issues

**Comment by KSR4599** [comment]
Yes, posted an issue.

**Comment by rweigel** [comment]
Great. Thanks.

The first thing that they are going to wonder is if it happens in more recent versions of Node. They won't want to fix it if it is due to a non-supported version of nodejs. When I write an issue, I try to be as comprehensive and helpful as possible and anticipate the first follow-up question that they will ask. They will probably also ask if or stand-alone commands such as `dir` work. The more helpful you are, the more likely they are to fix it or respond.

For my reference: https://github.com/mathisonian/command-exists/issues/30

**Comment by KSR4599** [comment]
Sure, I will be watchful about any possible follow-up questions.

**Comment by rweigel** [comment]
It is possible to edit what you posted to indicate if this happens in more recent versions of nodejs and if happens on other stand-alone commands such as dir. I recommend editing what you wrote.

**Comment by KSR4599** [comment]
Yes, professor, I have edited the content as specified. The issue persists even with the recent versions of the node (including the latest LTS v14.15.0). Like I mentioned in the issue, remaining standalone commands such as `dir, ls, pwd`.. seems to be recognized fine.

**Comment by rweigel** [comment]
At https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/cd I don't see that `cd` alone is a valid command as it is in the Unix shell or perhaps in PowerShell. Does `cd` in `cmd.exe` print an error? I think the issue may be that `cd` is not a valid command in the shell that is used by command-exists. If this is the case, it is not a bug in command-exists.

Does `cd C:\` work return `true`? I think this is what you want to test. I don't ever anticipate using the command `cd` alone.

**Comment by rweigel** [comment]
If `cd` is a valid command and `cd C:\` does not work either, then this

https://github.com/mathisonian/command-exists/blob/master/lib/command-exists.js#L96

should explain the issue.

**Comment by rweigel** [comment]
Didn't we discuss a while back that the issue in Windows was a compound command separated by a semicolon such as `cd /tmp; node SSCWeb2HAPI.js`? I thought we had a fix for that such as using `&&` instead of `;`. It seem like we have solved this problem before. Maybe look back at the repository history and our emails.

**Comment by KSR4599** [comment]
At [cd](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/cd), it says _when used without parameters, cd displays the current drive and directory_.  Likely, `cd` in `cmd.exe` is displaying the current directory.  I have also just tested out `cd C:\`  and it returns false.

**Comment by rweigel** [comment]
OK. Then I guess we leave the bug report as-is and continue with the switch of the check order if that will work.

**Comment by KSR4599** [comment]
> Didn't we discuss a while back that the issue in Windows was a compound command separated by a semicolon such as `cd /tmp; node SSCWeb2HAPI.js`? I thought we had a fix for that such as using `&&` instead of `;`. It seem like we have solved this problem before. Maybe look back at the repository history and our emails.

Yes, Professor, we have already fixed this. Yet, this [string](https://github.com/hapi-server/server-nodejs/blob/master/metadata/SSCWeb.json#L16) when we [split](https://github.com/hapi-server/server-nodejs/blob/4bf734b8e00632f81b73ab9a3653944959f8c4a4/lib/metadata.js#L245) based on spaces, the first chunk will result in  `cd`, which the command-exists is denying as a valid command.

**Comment by KSR4599** [comment]
> OK. Then I guess we leave the bug report as-is and continue with the switch of the check order if that will work.

Okay, I will switch the order and let you know if it works out.

**Comment by rweigel** [comment]
I just noted another potential problem with the approach in the old code at https://github.com/hapi-server/server-nodejs/blob/4bf734b8e00632f81b73ab9a3653944959f8c4a4/lib/metadata.js#L245.

If the command has leading whitespace, e.g., `str="    cd /tmp"`, the command to be checked is an empty string.

**Comment by KSR4599** [comment]
> I just noted another potential problem with the approach in the old code at
> 
> https://github.com/hapi-server/server-nodejs/blob/4bf734b8e00632f81b73ab9a3653944959f8c4a4/lib/metadata.js#L245
> 
> .
> If the command has leading whitespace, e.g., `str=" cd /tmp"`, the command to be checked is an empty string.

Yes. With the current code in-place, since white-space is not a valid command, it would automatically be treated as a file and starts [reading](https://github.com/hapi-server/server-nodejs/blob/4bf734b8e00632f81b73ab9a3653944959f8c4a4/lib/metadata.js#L265) it (which eventually gets failed).

**Comment by rweigel** [comment]
I think the reason that I was checking if the command was executable first was to account for something like

```
command="/tmp/path/to/executable"
```

In this case, `fs.exists` would return true and an attempt would be made to read and parse it as JSON. But in this case, the user has given a command that generates metadata in a single executable file that is not called by another executable. I should have stated this in the comments.  In the code, please put a link to this issue for the to-be-deprecated code that deals with a `command` node.

**Comment by rweigel** [comment]
For now, I think that it is fine to reverse the order. This will allow all of our old metadata to run without error because there were not cases that I recall where there was a command that was a single executable.

**Comment by KSR4599** [comment]
> For now, I think that it is fine to reverse the order. This will allow all of our old metadata to run without error because there were not cases that I recall where there was a command that was a single executable.

I have reversed the order and everything seems to work fine. I will create an associated test to make sure it is working with the old metadata and give you an update on it.

**Comment by rweigel** [comment]
Great. Go ahead and push the branch.

**Comment by rweigel** [comment]
Closed by 1e032e1
